app/assets/javascripts/discourse/app/services/presence.js

```diff
@@ -244,6 +233,65 @@ export default class PresenceService extends Service {
     });
   }
 
+  _getInitialData(channelName) {
+    let promiseProxy = this._initialDataRequests[channelName];
+    if (!promiseProxy) {
+      promiseProxy = this._initialDataRequests[
+        channelName
+      ] = createPromiseProxy();
+    }
+
+    once(this, this._makeInitialDataRequest);
+
+    return promiseProxy.promise;
+  }
+
+  async _makeInitialDataRequest() {
+    if (this._initialDataAjax) {
+      // try again next runloop
+      next(this, () => once(this, this._makeInitialDataRequest));
+    }
+
+    if (Object.keys(this._initialDataRequests).length === 0) {
+      // Nothing to request
+      return;
+    }
+
+    this._initialDataAjax = ajax("/presence/get", {
+      data: {
+        channels: Object.keys(this._initialDataRequests).slice(0, 50),
+      },
+    });
+
+    let result;
+    try {
+      result = await this._initialDataAjax;
+    } catch (e) {
+      this._initialDataAjax = null;
+      later(this, this._makeInitialDataRequest, PRESENCE_GET_RETRY_MS);
```

Comment:
Do we want to retry unconditionally here? Are there only a few specific errors we want to retry on?

Reply:
Yes I think we do. If this fails for any reason, the application will end up in a broken state until a refresh. Better to keep retrying every 5 seconds until the problem is resolved (e.g. the network connection is restored)